### ISSUE TITLE
Interpretation of Unicode characters while writing server.properties

### ISSUE THREAD

**Issue by KS-HTK** [itzg/docker-minecraft-server#1025]
minecrafts server.properties have a problem where setting the motd to `§6◕§5Test` gets translated to  `\u00C2\u00A76\u00E2\u0097\u0095\u00C2\u00A75Test` in the file after a single start.
I tried to prevent this by setting it to `\u00A76\u25D5\u00A75Test` in the docker-compose.yml but it seems to get interpreted while being written to the server.properties file resulting in the same `\u00C2\u00A76\u00E2\u0097\u0095\u00C2\u00A75Test` motd after server start.

It would be great if the Unicode strings could be added to the server.properties file as such and not as symbols.

**Comment by itzg** [comment]
This might be beyond the control of the image scripts, but to clarify how are you setting the motd value? Via image environment variable or directly in the file?

**Comment by KS-HTK** [comment]
I have just found out that the JDK actually contains a commandline tool to convert files to the correct encoding.
`native2ascii -encoding utf8 input.txt output.txt`
https://docs.oracle.com/javase/1.5.0/docs/tooldocs/windows/native2ascii.html

Issue: The command seems to have been removed at some point.
Tested it on the openjdk versions i have installed locally (Archlinux):
openjdk 7/8 present
openjdk 10/11/16 not present

**Comment by itzg** [comment]
These solutions seem to be beyond the scope of what this image should handle. Can you summarize something I could add to docs? Or you can PR a docs change.

**Comment by KS-HTK** [comment]
I do not quite understand what it is you want to have in the docs.
The problem is that this is not a step a user can do manually at first launch as it requires to be run after the `server.properties` file has been written but before the server launches.

For all versions before JDK9:
---
I am not very familiar with bash/shell scripting, but for everything using JDK8 or lower a single command can do the task:
`native2ascii -encoding utf8 server.properties server.properties`

For all versions since JDK9:
---

This is where it would require an Internet connection and compiling the downloaded files. I am not sure how one would check for failed tasks in a script. Here are the commands it would need get the tool:
```bash
curl http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/archive/235d2e871a2e.tar.gz/src/share/classes/sun/tools/native2ascii/ -o n2a.tar.gz
tar -xf n2a.tar.gz
mkdir build
javac -d ./build/ $(find jdk-235d2e871a2e/src/* | grep .java)
jar cfe n2a.jar sun/tools/native2ascii/Main -C build .
rm -r build/
rm -r jdk-235d2e871a2e
rm n2a.tar.gz
java -jar n2a.jar -encoding utf8 server.properties server.properties
```
As I am not familiar with shell scripting I don't really know how to write this as a useable script.
I am sure it needs to be a little more than just the commands.
I can understand if this is a litte to much to ask just to fix a bug mojang has not. 
___

For all others useing UTF8 Symbols (Formatting Codes like §a to §m) in there server.properties file I would recommend:
- start the server
- shut down the server
- change the server.properties file (with \udddd codes instead of the symbols)
- restart the server

and dont define the server.properties in the docker-compose

**Comment by itzg** [comment]
I guess we're both confused 😀

Thank you for a thorough explanation of what could be done to script this; however, there was an interesting thing you pointed out at the end:

> \udddd codes instead of the symbols

My expectation was that users would provide the escaped Unicode instead of the UTF8 symbols. I know that's an annoyance but the procedure you described is not trivial to add either.

**Comment by itzg** [comment]
Hang on...I went back to your original issue description and I'm now questioning if this is a problem at all. I certainly agree it is annoying to see escape codes in the properties file; however, were you finding the MOTD wasn't rendered correctly?

**Comment by KS-HTK** [comment]
>
> My expectation was that users would provide the escaped Unicode instead of the UTF8 symbols. I know that's an annoyance but the procedure you described is not trivial to add either.

It does not matter escaped or symbols in the docker compose are added to the server.properies file as Unicode symbols. This is the whole point of the issue. The minecraft server then converts these wrong (which is the minecraft issue) thus causing the annoyance.

> 
> Hang on...I went back to your original issue description and I'm now questioning if this is a problem at all. I certainly agree it is annoying to see escape codes in the properties file; however, were you finding the MOTD wasn't rendered correctly?

Technically yes.
If I set the motd in the server.properties to:
- `§bTest` the server then convets it to `\u00C2\u00A7bTest` (Renders Incorrectly)
- `\u00A7bTest` renders correctly
Setting in docker-compose:
- `§bTest` the server then convets it to `\u00C2\u00A7bTest` (Renders Incorrectly / expected)
- `\u00A7bTest` renders the same as `§bTest` (the reason for the bug-report).
This is what led me to beleave that the image does not write `\u00A7bTest` to the file but `§bTest`

**Comment by KS-HTK** [comment]
Just some additional Test:
To see if my assumption that the image adds Unicode as symbols and not as the escape sequence is true I have just done this:
- create a test/data/ folder for the image to use
- create a server.jar (empty file) in said data folder
- use this docker-compose:
```yml
version: '3.6'
services:
  minecraft:
    container_name: test
    image: itzg/minecraft-server
    environment:
      EULA: "TRUE"
      SERVER_NAME: "§6Test Server"
      MOTD: "\u00A76Test"
      OVERRIDE_SERVER_PROPERTIES: "true"

      TYPE: "CUSTOM"
      CUSTOM_SERVER: "server.jar"
    volumes:
      - "~/test/data:/data"
  ```
  
  This will obviously crash as the .jar file is corrupt. But it gives us a chance to see what the MOTD has been set to by the image.
  As I suspected before the MOTD in data/server.properties is set to `§6Test` which is not the defined MOTD and causes the issue.

**Comment by itzg** [comment]
It looks like `iconv` can be used for characters set conversions and that is already installed in the image. I'll do some research with that.

**Comment by KS-HTK** [comment]
I had tried to use `iconv` prior to suggesting `native2ascii`. `iconv` is only capable of performing the task if GNU's `libiconv` is present. For my system this is not the case.
The command I had found for this comes from [this](https://stackoverflow.com/a/40116175) stackoverflow answer.
`iconv -t ASCII --unicode-subst='\u%04X'`
I am not sure if `iconv` can perform this task without the lib at all. 
___
While searching for the `iconv` command I came across an other possibility using `jq`.
I do not know how to check if `jq` is in the image, but if it is here is how it could be applied:
```bash
motd="§6Test \u00a7bString"
tmp=$(echo \"${motd}\" | jq -a)
tmp=${tmp#\"}
tmp=${tmp%\"}
echo motd=$tmp >> server.properties
```
this would write the motd to the file using only escape sequences.
As I understand it this only converts non ascii symbols to escape sequences. It should not change any ascii characters.
So running it on all server.properties values should not cause a problem.
___
Playing around in the command line made me wonder how the escape sequences even get converted to symbols in the first place? Shouldn't it be possible to get the string as it is written in the docker-compose without any interpretation of the sequences?

**Comment by itzg** [comment]
Pull the latest image and give that a try. I created [a really small utility](https://github.com/itzg/asciify) to do the conversion exactly the way the script needed and now it processes the given `MOTD`.

**Comment by KS-HTK** [comment]
Ok, thanks.
Works as intended.

Just FYI, the issue applies to all values in server.properties
I guess the motd is the most affected because it is the only one where formatting makes sence but this also is an issue with non ascii language symbols. So it might be good to apply the fix to the server name.
Also the issue of \udddd escape strings being converted to chars actually still exsists for server name.

All fields that might be affected should be server-name, motd and possiibly resource-pack / texture-pack.
All others like level-name would probably cause other issues.
I'd rather the fix was applied to all string type fields. But the issue is resolved for me.

**Comment by itzg** [comment]
Yeah, I was trying to decide to process the whole file or per field. I'll continue with per field but great point about server name and others.
